context/app/static/js/components/workspaces/api.ts

```diff
@@ -0,0 +1,316 @@
+import useSWR from 'swr';
+import useSWRMutation from 'swr/mutation';
+
+import { useCallback } from 'react';
+
+import { useAppContext } from '../Contexts';
+import { Workspace, WorkspaceAPIResponse, WorkspaceJob } from './types';
+import { getWorkspaceHeaders } from './utils';
+
+/**
+ * Generates API URLs for various workspaces API actions
+ * @param workspacesEndpoint endpoint to use as the base for API URLs
+ * @returns an API URL generator
+ */
+export const apiUrls = (workspacesEndpoint: string) => ({
+  get jobTypes(): string {
+    return `${workspacesEndpoint}/job_types`;
+  },
+  get workspaces(): string {
+    return `${workspacesEndpoint}/workspaces`;
+  },
+  get jobs(): string {
+    return `${workspacesEndpoint}/jobs`;
+  },
+  delete(workspaceId: number): string {
+    return `${workspacesEndpoint}/workspaces/${workspaceId}`;
+  },
+  createWorkspaceFromNotebookPath(path: string): string {
+    return `/notebooks/${path}`;
+  },
+  get createWorkspaceFromTemplates(): string {
+    return `${workspacesEndpoint}/workspaces/`;
+  },
+  startWorkspace(workspaceId: number): string {
+    return `${workspacesEndpoint}/workspaces/${workspaceId}/start`;
+  },
+  stopJob(jobId: number): string {
+    return `${workspacesEndpoint}/jobs/${jobId}/stop`;
+  },
+});
+
+export function useWorkspacesApiURLs() {
+  const { workspacesEndpoint } = useAppContext();
+  return apiUrls(workspacesEndpoint);
+}
+
+export function useWorkspaceHeaders(): HeadersInit {
+  const { workspacesToken } = useAppContext();
+  return getWorkspaceHeaders(workspacesToken);
+}
+
+interface APIAction {
+  url: string;
+  headers?: HeadersInit;
+}
+
+interface StopJobArgs extends APIAction {
+  jobId: number;
+}
+
+/**
+ * Stop a currently-running workspace job
+ * @param key Readable key for the current action
+ * @param param2.arg.headers The headers to use for the request
+ * @param param2.arg.jobId ID of the job to stop
+ * @param param2.arg.url The URL to use for the request
+ * @returns The response from the server if the request was successful
+ */
+async function fetchStopJob(_key: string, { arg: { headers, jobId, url } }: { arg: StopJobArgs }) {
+  const response = await fetch(url, {
+    method: 'PUT',
+    headers,
+  });
+  if (!response.ok) {
+    throw Error(`Job stop for job ${jobId} failed`);
+  }
+  return response;
+}
+
+export function useStopJob() {
+  const apiURLs = useWorkspacesApiURLs();
+  const headers = useWorkspaceHeaders();
+
+  const { trigger, isMutating } = useSWRMutation('stop-job', fetchStopJob);
+
+  const stopJobCallback = useCallback(
+    async (jobId: number) => {
+      await trigger({
+        jobId,
+        url: apiURLs.stopJob(jobId),
+        headers,
+      });
+    },
+    [apiURLs, headers, trigger],
+  );
```

Comment:
Pull `apiURLs` and `headers` into the useCallback so they don't have to be dependencies?

Reply:
Both `apiURLs` and `headers` are provided by hooks which take care of providing the appropriate `AppContext` values (i.e. the workspaces endpoint and token.) Pulling them into the `useCallback` would result in those AppContext values being dependencies instead.